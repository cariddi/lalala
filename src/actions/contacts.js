
import { LIST_CONTACT_URL } from '../config/urls';
import { FETCH_CONTACTS } from './types';
	
export const fetchContacts = () => ( dispatch, _, api ) => dispatch( {
	type: FETCH_CONTACTS,
	promise: api
		.get( `${LIST_CONTACT_URL}` )
		.then( response => response.response )
	
} );