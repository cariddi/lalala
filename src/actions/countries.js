import { FETCH_COUNTRIES } from './types';
import { LIST_COUNTRIES_URL } from '../config/urls';

export const fetchCountries = () => ( dispatch, _, api ) => dispatch( {
	type: FETCH_COUNTRIES,
	promise: api
		.get( `${LIST_COUNTRIES_URL}` )
		.then( response => response.response )
} );
