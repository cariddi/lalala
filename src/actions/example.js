import { types } from './types';

export function exampleAction() {
	return {
		type: types.EXAMPLE,
		promise: Promise.resolve(  )
	};
}
