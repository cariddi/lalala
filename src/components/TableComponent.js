import React from 'react'
import { Table } from 'semantic-ui-react'
import { Link } from 'react-router-dom'

const TableComponent = ( { contacts } ) => {
    if( !(contacts && contacts.length) ) return <h3>No contact list cat</h3>;

    return (
        <div className="mainContainer">
            <div className="mainHeader">
                <div className="avatar">User Avatar</div>
                <h2 className="headerTitle">User Name</h2>
                <button className="button">Add Contact</button>
            </div>
            <hr />
            <div className="dataContainer">
                <div className="subHeader">
                    <div className="subHeaderLabel">Total: {contacts.length}</div>
                    <h2 className="subHeaderTitle"><strong>Contacts</strong></h2>
                </div>
                <br />
                <Table celled>
                    <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell className="tableHeaderCell">ID</Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell">Avatar</Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell">Name</Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell">Email</Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell">Phone1</Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell">Phone2</Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell">Address</Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell">Birthdate</Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell"></Table.HeaderCell>
                            <Table.HeaderCell className="tableHeaderCell"></Table.HeaderCell>
                        </Table.Row>
                    </Table.Header>
                    <Table.Body>
                        {
                            contacts.map(c => {
                                return (
                                    <Table.Row key={c.id} className="tableRow">
                                        <Table.Cell className="tableCell">{c.id}</Table.Cell>
                                        <Table.Cell className="tableCell">{c.avatar}</Table.Cell>
                                        <Table.Cell className="tableCell">{c.name}</Table.Cell>
                                        <Table.Cell className="tableCell">{c.email}</Table.Cell>
                                        <Table.Cell className="tableCell">{c.phone1}</Table.Cell>
                                        <Table.Cell className="tableCell">{c.phone2}</Table.Cell>
                                        <Table.Cell className="tableCell">{c.address}</Table.Cell>
                                        <Table.Cell className="tableCell">{c.birthdate}</Table.Cell>
                                        <Table.Cell className="tableCell"><Link to={`/edit/${c.id}`}><button className="linkBtnEdit">Edit</button></Link></Table.Cell>
                                        <Table.Cell className="tableCell"><Link to={`/view/${c.id}`}><button className="linkBtnView">+</button></Link></Table.Cell>
                                    </Table.Row>
                                )
                                }
                            )
                        }
                    </Table.Body>
                </Table>
            </div>
        </div>
    )
}

export default TableComponent
