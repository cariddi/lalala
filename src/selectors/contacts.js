import { createSelector } from 'reselect';

const getContacts = state => state.contacts;
const getContactData = ( state, props ) => {
    return props.match.params.id;
};

export const getContactsSelector = createSelector(
    getContacts,
    contacts => contacts.contacts
);

export const getContactDataSelector = createSelector(
    [ getContacts, getContactData ],
    ( contacts, contactID) => {
        if( !(contacts.contacts && contacts.contacts.length) ) return;
        
        return contacts.contacts.filter( c => c.id == contactID )[ 0 ];
    }
);

// let fakeState = {
//     contacts: [
//         {
//             "id": 5,
//             "name": "Susana Córdova Pabón",
//             "email": "ariadna_ruiz@abrego.es",
//             "phone1": "972853428",
//             "phone2": "631067997",
//             "birthdate": "1995-08-10",
//             "address": "Chalet Mónica Perea 7 Esc. 483",
//             "state": {
//                 "id": 29,
//                 "name": "Bahia"
//             },
//             "country": {
//                 "id": 3,
//                 "name": "Brasil"
//             },
//             "avatar": null
//         }
//     ]
// }

// console.log(getContactsKeysSelector(fakeState))