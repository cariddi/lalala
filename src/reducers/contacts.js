import { makeAsyncActionReducer, handleSuccess } from '../lib/reduxUtils';
import { FETCH_CONTACTS, FETCH_CONTACT_DATA, EDIT_CONTACT } from '../actions/types';

export const fetchContactsRequest = makeAsyncActionReducer( FETCH_CONTACTS );
export const fetchContactDataRequest = makeAsyncActionReducer( FETCH_CONTACT_DATA );
export const editContactRequest = makeAsyncActionReducer( EDIT_CONTACT );

export const contacts = ( state = {}, action ) => {
	switch ( action.type ) {
	case FETCH_CONTACTS:
		return handleSuccess( state, action, ( prevState ) => {
			const contacts = action.payload;
			return { ...prevState, contacts: contacts };
		} );
	case FETCH_CONTACT_DATA:
	case EDIT_CONTACT:
		return handleSuccess( state, action, ( prevState ) => {
			const contactData = action.payload;
			console.log('EDIT: ', contactData);
			if( !contactData ) return { ...prevState } ;
			if( !prevState.contacts ) prevState.contacts = [];
			if( !prevState.contacts.includes(contactData) ) prevState.contacts.push(contactData);
			
			return { ...prevState, contactID: contactData.id };
		} );
	default:
		return state;
	}
};
