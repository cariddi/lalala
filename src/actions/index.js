import * as contactsActions from './contacts';
import * as countriesActions from './countries';
import * as statesActions from './states';

export default {
	...contactsActions,
	...countriesActions,
	...statesActions
};
