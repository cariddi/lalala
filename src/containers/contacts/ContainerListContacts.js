import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connectComponent } from '../../lib/connectComponent';
import { getContacts } from '../../selectors/contacts';

export class ContainerListContacts extends Component {
    componentWillMount() {
		if ( !this.contacts ) { this.fetchContacts(); }
	}

    get contacts() { return this.props.contacts; }

    fetchContacts() { this.props.fetchContacts( ); }

    render() {
        const { contacts } = this;

        return (
            <div className='ListContacts'>
                Contact List
                {
                    this.props.contacts ?

                    this.props.contacts.map((c, idx) => {
                        return (
                            <li key={idx}>{c.id - c.name}</li>
                        )
                    })
                    :
                    <p>No contacts</p>
                }
            </div>
        )
    }
}

ContainerListContacts.propTypes = {
	contacts: PropTypes.any,
	fetchContacts: PropTypes.func.isRequired
};

ContainerListContacts.defaultProps = {
	contacts: undefined
};

export default connectComponent( ( state, props ) => ( {
	fetchContactsRequest: state.fetchContactsRequest, //reducer
	contacts: getContacts( state, props ) //selector
} ) )( ContainerListContacts );
