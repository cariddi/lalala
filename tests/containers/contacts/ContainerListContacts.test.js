import React from 'react';
import { shallow } from 'enzyme';
import { ContainerListContacts } from '../../../src/containers/contacts/ContainerListContacts';
import TableComponent from '../../../src/components/TableComponent';
import { fetchContactData } from '../../../src/actions/contacts';

describe( 'ContactsList Container', () => {
	const fetchContacts = jest.fn();
	let container;

	beforeEach( () => {
		container = shallow(
			<ContainerListContacts 
				fetchContacts={fetchContacts}
			/>
		)
	} );

	it( 'passes the contacts list to the Table component', () => {
		expect( container.find( TableComponent ) ).not.toBeNull();
		expect( container ).toMatchSnapshot();
	} );

	// describe( 'shows the contact details when clicked', () => {
	// 	const contactID = 1;
	// 	const fetchContactDataMocked = jest.fn()
	// 		.mockImplementation( fetchContactData => fetchContactData(1) );

	// 	beforeEach( () => {
	// 		container.find( TableComponent ).simulate( 'click', contactID );
	// 	} );

	// 	it( 'calls the fetchContactData prop with the contact ID', () => {
	// 		expect( fetchContactDataMocked ).toHaveBeenCalledWith( contactID );
	// 	} );
	// } );
} );
