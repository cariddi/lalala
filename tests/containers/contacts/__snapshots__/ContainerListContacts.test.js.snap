// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsList Container passes the contacts list to the Table component 1`] = `
<div
  className="ListContacts"
>
  <h2>
    Contact List
  </h2>
  <TableComponent />
</div>
`;
